File: src/org.eclipse.ice.reactor/src/org/eclipse/ice/reactor/pwr/ControlBank.java
/*******************************************************************************
 * Copyright (c) 2012, 2014 UT-Battelle, LLC.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *   Initial API and implementation and/or initial documentation - Jay Jay Billings,
 *   Jordan H. Deyton, Dasha Gorin, Alexander J. McCaskey, Taylor Patterson,
 *   Claire Saunders, Matthew Wang, Anna Wojtowicz
 *******************************************************************************/
package org.eclipse.ice.reactor.pwr;

import ncsa.hdf.object.h5.H5File;
import ncsa.hdf.object.h5.H5Group;

import org.eclipse.ice.io.hdf.HdfReaderFactory;
import org.eclipse.ice.io.hdf.HdfWriterFactory;
import org.eclipse.ice.reactor.HDF5LWRTagType;
import org.eclipse.ice.reactor.ILWRComponentVisitor;
import org.eclipse.ice.reactor.LWRComponent;

/**
 * <p>
 * The ControlBank class contains properties associated with axially movable
 * RodClusterAssembly objects known as Rod Cluster Control Assemblies. A Bank is
 * a group of RCCA's that are positioned simultaneously by the plant operations.
 * A Bank may have 8 individual RCCAs, for instance, located symmetrically
 * around the core. For a core with 50+ RCCAs, usually only 8 or so banks are
 * controlled independently.
 * </p>
 * 
 * @author Scott Forest Hull II
 */
public class ControlBank extends LWRComponent {
	/**
	 * <p>
	 * The distance between an axial step.
	 * </p>
	 * 
	 */
	private double stepSize;
	/**
	 * <p>
	 * The maximum number of axial steps.
	 * </p>
	 * 
	 */
	private int maxNumberOfSteps;

	/**
	 * <p>
	 * The nullary Constructor.
	 * </p>
	 * 
	 */
	public ControlBank() {
		this.setMaxNumberOfSteps(1);
		this.setStepSize(0.0);
		this.setName("ControlBank 1");
		this.setDescription("Default Control Bank");
		this.setId(1);
		this.HDF5LWRTag = HDF5LWRTagType.CONTROL_BANK;

	}

	/**
	 * <p>
	 * A parameterized Constructor.
	 * </p>
	 * 
	 * @param name
	 *            The name of the component.
	 * @param stepSize
	 *            <p>
	 *            The distance between an axial step.
	 *            </p>
	 * @param maxNumberOfSteps
	 *            <p>
	 *            The maximum number of axial steps.
	 *            </p>
	 */
	public ControlBank(String name, double stepSize, int maxNumberOfSteps) {
		this();
		this.setStepSize(stepSize);
		this.setMaxNumberOfSteps(maxNumberOfSteps);
		this.setName(name);
	}

	/**
	 * <p>
	 * Sets the axial step size.
	 * </p>
	 * 
	 * @param stepSize
	 *            <p>
	 *            The axial step size.
	 *            </p>
	 */
	public void setStepSize(double stepSize) {
		if (stepSize > 0.0) {
			this.stepSize = stepSize;
		} else {
			this.stepSize = 0.0;
		}
	}

	/**
	 * <p>
	 * Sets the maximum number of axial steps.
	 * </p>
	 * 
	 * @param maxNumberOfSteps
	 *            <p>
	 *            The maximum number of axial steps.
	 *            </p>
	 */
	public void setMaxNumberOfSteps(int maxNumberOfSteps) {
		if (maxNumberOfSteps > 0) {
			this.maxNumberOfSteps = maxNumberOfSteps;
		} else {
			this.maxNumberOfSteps = 1;
		}
	}

	/**
	 * <p>
	 * Returns the maximum number of axial steps.
	 * </p>
	 * 
	 * @return <p>
	 *         The maximum number of axial steps.
	 *         </p>
	 */
	public int getMaxNumberOfSteps() {
		return this.maxNumberOfSteps;
	}

	/**
	 * <p>
	 * Returns the axial step size.
	 * </p>
	 * 
	 * @return <p>
	 *         The axial step size.
	 *         </p>
	 */
	public double getStepSize() {
		return this.stepSize;
	}

	/**
	 * <p>
	 * Calculates and returns the stroke length, which is the axial step size
	 * multiplied by the maximum number of axial steps.
	 * </p>
	 * 
	 * @return <p>
	 *         The axial step size multiplied by the maximum number of axial
	 *         steps.
	 *         </p>
	 */
	public double getStrokeLength() {

		return ((double) this.maxNumberOfSteps) * this.stepSize;
	}

	/*
	 * Overrides a method from LWRComponent.
	 */
	@Override
	public boolean writeAttributes(H5File h5File, H5Group h5Group) {
		boolean flag = true;

		flag &= super.writeAttributes(h5File, h5Group);
		flag &= HdfWriterFactory.writeDoubleAttribute(h5File, h5Group,
				"stepSize", stepSize);
		flag &= HdfWriterFactory.writeIntegerAttribute(h5File, h5Group,
				"maxNumberOfSteps", maxNumberOfSteps);

		return flag;
	}

	/*
	 * Overrides a method from LWRComponent.
	 */
	@Override
	public boolean equals(Object otherObject) {

		// Local Declarations
		boolean retVal = false;
		ControlBank bank;

		// If they are the same on the heap, return true
		if (otherObject == this) {
			return true;
		}

		// Make sure the otherObject is not null and an instance of this object
		if (otherObject != null && otherObject instanceof ControlBank) {

			// Cast it
			bank = (ControlBank) otherObject;

			// Compare values
			retVal = (super.equals(otherObject)
					&& this.maxNumberOfSteps == bank.maxNumberOfSteps && this.stepSize == bank.stepSize);
		}

		// Return retVal
		return retVal;

	}

	/*
	 * Overrides a method from LWRComponent.
	 */
	@Override
	public int hashCode() {

		// Local Declarations
		int hash = super.hashCode();

		hash += 31 * this.maxNumberOfSteps;
		hash += 31 * this.stepSize;

		return hash;

	}

	/**
	 * <p>
	 * Deep copies the contents of the object.
	 * </p>
	 * 
	 * @param otherObject
	 *            <p>
	 *            The object to be copied.
	 *            </p>
	 */
	public void copy(ControlBank otherObject) {

		// If otherObject is null, return
		if (otherObject == null) {
			return;
		}

		// Copy contents - super
		super.copy(otherObject);

		// Copy local contents
		this.maxNumberOfSteps = otherObject.maxNumberOfSteps;
		this.stepSize = otherObject.stepSize;

	}

	/*
	 * Overrides a method from LWRComponent.
	 */
	@Override
	public Object clone() {

		// Local Declarations
		ControlBank bank = new ControlBank();

		// Copy contents
		bank.copy(this);

		// Return newly instantiated object
		return bank;

	}

	/*
	 * Overrides a method from LWRComponent.
	 */
	@Override
	public boolean readAttributes(H5Group h5Group) {

		// Local Declarations
		boolean flag = true;

		// Get values
		Double stepSize = HdfReaderFactory.readDoubleAttribute(h5Group,
				"stepSize");
		Integer maxNumberOfSteps = HdfReaderFactory.readIntegerAttribute(
				h5Group, "maxNumberOfSteps");

		// Call super
		flag &= super.readAttributes(h5Group);

		// check values
		if (stepSize == null || maxNumberOfSteps == null || !flag
				|| h5Group == null) {
			return false;
		}

		// If everything is valid, then set data
		this.stepSize = stepSize.doubleValue();
		this.maxNumberOfSteps = maxNumberOfSteps.intValue();

		return true;

	}

	/*
	 * Overrides a method from LWRComponent.
	 */
	@Override
	public void accept(ILWRComponentVisitor visitor) {
		visitor.visit(this);
	}
}

File: src/native/org.eclipse.ice.reactor.native/pwr/ControlBank.cpp
/*******************************************************************************
* Copyright (c) 2012, 2014 UT-Battelle, LLC.
* All rights reserved. This program and the accompanying materials
* are made available under the terms of the Eclipse Public License v1.0
* which accompanies this distribution, and is available at
* http://www.eclipse.org/legal/epl-v10.html
*
* Contributors:
*   Initial API and implementation and/or initial documentation - Jay Jay Billings,
*   Jordan H. Deyton, Dasha Gorin, Alexander J. McCaskey, Taylor Patterson,
*   Claire Saunders, Matthew Wang, Anna Wojtowicz
*******************************************************************************/

#include "ControlBank.h"
#include "../HDF5LWRTagType.h"
#include <string>
#include <vector>
#include <HdfReaderFactory.h>
#include <HdfWriterFactory.h>
#include <H5Cpp.h>
#include "../LWRComponent.h"

using namespace ICE_Reactor;

/**
 * The Copy Constructor
 */
ControlBank::ControlBank(ControlBank & arg) : LWRComponent(arg) {

    // begin-user-code

    this->stepSize = arg.stepSize;
    this->maxNumberOfSteps = arg.maxNumberOfSteps;
    // end-user-code

}

/**
 * The Destructor
 */
ControlBank::~ControlBank() {
    //TODO Auto-generated method stub
}

/**
 * The nullary Constructor.
 */
ControlBank::ControlBank() {

    // begin-user-code

    ControlBank::setMaxNumberOfSteps(1);
    ControlBank::setStepSize(0.0);
    ControlBank::setName("ControlBank 1");
    ControlBank::setDescription("Default Control Bank");
    ControlBank::setId(1);
    this->HDF5LWRTag = CONTROL_BANK;

    // end-user-code
}

/**
 * A parameterized Constructor.
 *
 * @param the name
 * @param the stepSize
 * @param the maxNumberOfSteps
 */
ControlBank::ControlBank(const std::string name, double stepSize, int maxNumberOfSteps) {

    // begin-user-code
    this->name = "ControlBank 1";
    ControlBank::setMaxNumberOfSteps(maxNumberOfSteps);
    ControlBank::setStepSize(stepSize);
    ControlBank::setName(name);
    ControlBank::setDescription("Default Control Bank");
    ControlBank::setId(1);
    this->HDF5LWRTag = CONTROL_BANK;

    // end-user-code

}

/**
 * Sets the axial step size.
 *
 * @param the stepSize
 */
void ControlBank::setStepSize(double stepSize) {

    // begin-user-code
    if (stepSize > 0.0) {
        this->stepSize = stepSize;
    } else {
        this->stepSize = 0.0;
    }

    return;

    // end-user-code

}

/**
 * Sets the maximum number of axial steps.
 *
 * @param the maxNumberOfSteps to set
 */
void ControlBank::setMaxNumberOfSteps(int maxNumberOfSteps) {

    // begin-user-code

    //IF the maxSteps passed are not correct, return
    if (maxNumberOfSteps > 0) {

        this->maxNumberOfSteps = maxNumberOfSteps;
    } else {

        this->maxNumberOfSteps = 1;
    }

    // end-user-code
    return;
}

/**
 * Returns the maximum number of axial steps.
 *
 * @return the maxNumberOfSteps
 */
int ControlBank::getMaxNumberOfSteps() {

    // begin-user-code

    return this->maxNumberOfSteps;

    // end-user-code
}

/**
 * Returns the axial step size.
 *
 * @return the stepSize
 */
double ControlBank::getStepSize() {

    // begin-user-code

    return this->stepSize;

    // end-user-code

}

/**
 * Calculates and returns the stroke length, which is the axial step size multiplied by the maximum number of axial steps.
 *
 * @return the stroke length
 */
double ControlBank::getStrokeLength() {

    //begin-user-code

    return ((double) this->maxNumberOfSteps) * this->stepSize;

    // end-user-code
}

/**
 * Overrides the equals operation to check the attributes on this object with another object of the same type.  Returns true if the objects are equal.  False otherwise.
 *
 * @param object to compare to
 *
 * @return true if equal, false otherwise
 */
bool ControlBank::operator==(const ControlBank& otherObject) const {

    return LWRComponent::operator ==(otherObject) && this->maxNumberOfSteps == otherObject.maxNumberOfSteps && this->stepSize == otherObject.stepSize;
}

/**
 * Deep copies and returns a newly instantiated object.
 *
 * @return the newly instantiated object
 */
std::shared_ptr<ICE_DS::Identifiable> ControlBank::clone() {

    // begin-user-code

    // Local Declarations
    std::shared_ptr<ControlBank> component(new ControlBank (*this));

    // Return the newly instantiated object
    return component;

    // end-user-code

}

/**
 * This operation writes HDF5 Attributes to the metadata of h5Group in the h5File. If the h5Group is null or h5File is null or can not be opened, then false is returned. If the operation fails to write all Attributes, then false is returned. Otherwise, true is returned.
 *
 * @param The File
 * @param The Group
 *
 * @return True if successful, false otherwise
 */
bool ControlBank::writeAttributes(std::shared_ptr<H5::H5File> h5File, std::shared_ptr<H5::Group> h5Group) {

    // begin-user-code

    bool flag = true;

    flag &= LWRComponent::writeAttributes(h5File, h5Group);
    flag &= ICE_IO::HdfWriterFactory::writeDoubleAttribute(h5File, h5Group, "stepSize", this->stepSize);
    flag &= ICE_IO::HdfWriterFactory::writeIntegerAttribute(h5File, h5Group, "maxNumberOfSteps", this->maxNumberOfSteps);

    return flag;


    // end-user-code

}

/**
 * This operation reads Attributes from h5Group and assigns their values to class variables. If h5Group is null, false is returned. If any Attribute values are null, false is returned. Otherwise, true is returned.
 *
 * @param  The Group
 *
 * @return True if successful, false otherwise
 */
bool ControlBank::readAttributes(std::shared_ptr<H5::Group> h5Group) {

    // begin-user-code

    //Local attributes (so we only call read ONCE)
    //These will clear out by the garbage collector
    double stepSize;
    int maxNumberOfSteps;

    //Check super
    if(h5Group.get() == NULL) {
        return false;
    }

    //Get the information.  If any fail out, return false and do not change data.
    try {
        if (LWRComponent::readAttributes(h5Group) == false) return false;
        stepSize = ICE_IO::HdfReaderFactory::readDoubleAttribute(h5Group, "stepSize");
        maxNumberOfSteps = ICE_IO::HdfReaderFactory::readIntegerAttribute(h5Group, "maxNumberOfSteps");
    } catch (...) {
        return false;
    }

    //Set the primitive data
    this->stepSize = stepSize;
    this->maxNumberOfSteps = maxNumberOfSteps;

    return true;

    // end-user-code
}
